modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java

```diff
@@ -0,0 +1,487 @@
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+
+                return failedFuture(e);
+            } catch (ExecutionException e) {
+                return failedFuture(e);
+            }
+
+            // Fire event early, because dependent listeners have to register VersionedValues' update futures
+            var eventParams = new SchemaEventParameters(causalityToken, tableId, newSchema);
+
+            fireEvent(SchemaEvent.CREATE, eventParams)
+                    .whenComplete((v, e) -> {
+                        if (e != null) {
+                            LOGGER.warn("Error when processing CREATE event", e);
+                        }
+                    });
+
+            return registriesVv.update(causalityToken, (registries, e) -> inBusyLock(busyLock, () -> {
+                if (e != null) {
+                    return failedFuture(new IgniteInternalException(IgniteStringFormatter.format(
+                            "Cannot create a schema for the table [tblId={}, ver={}]", tableId, newSchemaVersion), e)
+                    );
+                }
+
+                return registerSchema(registries, tableId, newSchema);
+            })).thenApply(ignored -> false);
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    private void setColumnMapping(SchemaDescriptor schema, int tableId) throws ExecutionException, InterruptedException {
+        if (schema.version() == CatalogTableDescriptor.INITIAL_TABLE_VERSION) {
+            return;
+        }
+
+        int prevVersion = schema.version() - 1;
+
+        SchemaDescriptor prevSchema = searchSchemaByVersion(tableId, prevVersion);
+
+        if (prevSchema == null) {
+            // This is intentionally a blocking call, because this method is used in a synchronous part of the configuration listener.
+            // See the call site for more details.
+            prevSchema = schemaByVersion(tableId, prevVersion).get();
+        }
+
+        schema.columnMapping(SchemaUtils.columnMapper(prevSchema, schema));
+    }
+
+    /**
+     * Registers the new schema in a Schema Registry.
+     *
+     * @param registries Map of schema registries.
+     * @param tableId Table id.
+     * @param schema Schema descriptor.
+     * @return Future that, when complete, will resolve into an updated map of schema registries
+     *     (to be used in {@link IncrementalVersionedValue#update}).
+     */
+    private CompletableFuture<Map<Integer, SchemaRegistryImpl>> registerSchema(
+            Map<Integer, SchemaRegistryImpl> registries,
+            int tableId,
+            SchemaDescriptor schema
+    ) {
+        ByteArray key = schemaWithVerHistKey(tableId, schema.version());
+
+        byte[] serializedSchema = SchemaSerializerImpl.INSTANCE.serialize(schema);
+
+        return metastorageMgr.invoke(notExists(key), put(key, serializedSchema), noop())
+                .thenApply(t -> {
```

Comment:
Let's move this to the outside, and rename method to smth like `saveSchema`.
Thus, the method will represent a single action, its name will corresponds the content, and will `registries` changes will be localized.